### DISCUSSION TITLE
Mandatory parameter for public instances! Tutorial

### DISCUSSION THREAD

**Discussion by unixfox** [searxng/searx-instances#417] [Announcements]
If you need help in configuring this new parameter please do not hesitate to ask for help:
- join our matrix channel: https://matrix.to/#/#searxng:matrix.org and say that you are running a public instance
- create a github discussion: https://github.com/searxng/searxng/discussions

This new parameter activates both the bot limiter AND the new advanced bot limiter called link_token. This is to improve the quality of the public instances by decreasing overall amount of bots harming the public instances.

It is very possible that in the future we will add more bot protection features enabled by default for you, but thanks to this parameter you won't have to do anything.

# Basic tutorial
**Only if you already have the limiter enabled, thus valkey activated and your web server correctly configured! If you do not please see [per type of installation](#per-type-of-installation).**

1. Update SearXNG to the latest version.
2. Go to your settings.yml file
3. Add the line `public_instance: true` in the section `server:`. Example:
```
server:
  public_instance: true
```
3. Restart SearXNG

# Per type of installation

- [searxng-docker](#searxng-docker)
- [installation script (./utils/searxng.sh)](#installation-script-utilssearxngsh)
- [Step by step installation (type) AND other types of custom installation (not Docker based)](#step-by-step-installation-type-and-other-types-of-custom-installation-not-docker-based)
- [Any other non official installation - Docker based](#any-other-non-official-installation---docker-based)

## searxng-docker
searxng-docker already have both the limiter and redis activated it's a simple change for you.

1. Go to the directory of searxng-docker
2. Update the directory by pulling the latest changes: `git pull`
3. Edit the file `settings.yml`.
4. Add the line `public_instance: true` in the section `server:`. Example:
```
server:
  public_instance: true
```
5. Update SearXNG and restart it

```
docker-compose pull
docker-compose up -d
docker-compose restart
```

## installation script (./utils/searxng.sh)
Depending on when you installed SearXNG you may already have the limiter and redis enabled.

### 1) Check if already have redis enabled or not
1. View the file `/etc/searxng/settings.yml` with cat for example.
2. If you already have `url: unix:///usr/local/searxng-redis/run/redis.sock?db=0` then follow [the first step](#b-if-you-do-not-have-redis), otherwise [the second one](#b-if-you-do-not-have-redis).

#### A) If you already have redis
1. Pull the latest SearXNG changes:

```
./utils/searxng.sh update
```
2. Edit the file `/etc/searxng/settings.yml`.
3. Add the line `public_instance: true` in the section `server:`. Example:
```
server:
  public_instance: true
```
4. Restart SearXNG:

```
service uwsgi restart searxng
```
Note: It may be another command but if you have installed SearXNG with uwsgi, it's this one.

#### B) If you do not have redis
1. Pull the latest SearXNG changes:

```
./utils/searxng.sh update
```
2. Install redis:

```
./utils/searxng.sh install redis
```
Accept all the steps, this may take a while.

3. Edit the file `/etc/searxng/settings.yml`.
4. Add the line `public_instance: true` in the section `server:`. Example:
```
server:
  public_instance: true
```
5. Restart SearXNG:

```
service uwsgi restart searxng
```
Note: It may be another command but if you have installed SearXNG with uwsgi, it's this one.

#### 3) Adapt your web server
- If you are using NGINX, check that you did correctly follow our guide: https://docs.searxng.org/admin/installation-nginx.html#installation-nginx
- If you are using apache, check that you did correctly follow our guide: https://docs.searxng.org/admin/installation-apache.html#installation-apache
- If you are using another web server, check that you are correctly sending the headers `X-Real-IP` and `X-Forwarded-For` to the SearXNG server. See the NGINX or apache tutorial for advice.

## Step by step installation (type) AND other types of custom installation (not Docker based)
You probably do not have redis enabled so you will have to configure redis.

### 1) Install Redis
1. Install redis with your package manager. On debian/ubuntu: `sudo apt install redis`
2. Edit the redis config file: `/etc/redis/redis.conf`
3. Check that you have redis running on localhost, check if the lines `bind 127.0.0.1 -::1` and `port 6379` exist, if not add them.
4. Restart redis if needed (above line modified or not).

### 2) Configure redis and new parameter in SearXNG
1. Update SearXNG to the latest version. Probably with `git pull`.
2. Edit your `settings.yml`. It is very probably at `/etc/searxng/settings.yml`.
3. Add the line `public_instance: true` in the section `server:`. Example:
```
server:
  public_instance: true
```
4. In the section `redis:`, add the line `url: redis://localhost:6379/1`:

```
redis:
  url: redis://localhost:6379/1
```

Please see the documentation for further details: https://docs.searxng.org/admin/settings/settings_redis.html#settings-redis

5. Restart SearXNG

### 3) Adapt your web server
The point is to have the headers `X-Real-IP` and `X-Forwarded-For` correctly sent to the SearXNG server.

- If you are using NGINX, check that you did correctly follow our guide: https://docs.searxng.org/admin/installation-nginx.html#installation-nginx
- If you are using apache, check that you did correctly follow our guide: https://docs.searxng.org/admin/installation-apache.html#installation-apache
- If you are using another web server, check that you are correctly sending the headers `X-Real-IP` and `X-Forwarded-For` to the SearXNG server. See the NGINX or apache tutorial for advice.

### 1) Redis and configure SearXNG
1. Install Redis locally. If you are on Docker add it for example in your docker-compose.yml file.

```
services:
  redis:
    image: redis:alpine
```
2. Update SearXNG to the latest version. Probably with `git pull` or `docker compose pull` then `docker compose up -d`.
3. Edit your `settings.yml`. It is very probably at `/etc/searxng/settings.yml`.
4. Add this line for the `server:` section:

```
server:
  public_instance: true
```
5. In the section `redis:`, add the line `url: redis://localhost:6379/1`:

```
redis:
  url: redis://localhost:6379/1
```

Please see the documentation for further details: https://docs.searxng.org/admin/settings/settings_redis.html#settings-redis

If you are on Docker, for redis use this line instead: `redis://redis:6379/1`

5. Restart SearXNG
### 2) Adapt your web server
The point is to have the headers `X-Real-IP` and `X-Forwarded-For` correctly sent to the SearXNG server.

- If you are using NGINX, check that you did correctly follow our guide: https://docs.searxng.org/admin/installation-nginx.html#installation-nginx
- If you are using apache, check that you did correctly follow our guide: https://docs.searxng.org/admin/installation-apache.html#installation-apache
- If you are using another web server, check that you are correctly sending the headers `X-Real-IP` and `X-Forwarded-For` to the SearXNG server. See the NGINX or apache tutorial for advice.

**Comment by unixfox** [comment]
@return42 could you turn on the parameter on your instance? you should already have everything setup (redis wise), but this way your instance is conformed like the others.

thank you

**Reply by return42** [reply]
Oops .. thanks for pointing out :+1: .. done:

![image](https://github.com/searxng/searx-instances/assets/554536/a39dddc3-c6a4-4a9a-b2ef-3071fb5e0cea)
